packages/eas-cli/src/update/utils.ts

```diff
@@ -269,3 +274,108 @@ export function getBranchDescription(branch: UpdateBranchFragment): FormattedBra
     },
   };
 }
+
+export function isBundleDiffingEnabled(exp: ExpoConfig): boolean {
+  return exp.updates?.enableBsdiffPatchSupport === true;
+}
+
+export function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
+  let timer: ReturnType<typeof setTimeout> | undefined;
+
+  return new Promise<T>((resolve, reject) => {
+    timer = setTimeout(() => {
+      reject(new Error(`Timed out after ${ms} ms`));
+    }, ms);
+
+    promise.then(resolve, reject).finally(() => {
+      if (timer) {
+        clearTimeout(timer);
+      }
+    });
+  });
+}
+
+// Make authenticated requests to the launch asset URL with diffing headers
+export async function prewarmDiffingAsync(
+  graphqlClient: ExpoGraphqlClient,
+  appId: string,
+  newUpdates: UpdatePublishMutation['updateBranch']['publishUpdateGroups']
+): Promise<void> {
+  const DUMMY_EMBEDDED_UPDATE_ID = '00000000-0000-0000-0000-000000000000';
+
+  const toPrewarm = [] as {
+    update: UpdatePublishMutation['updateBranch']['publishUpdateGroups'][0];
+    launchAssetKey: string;
+  }[];
+
+  for (const update of newUpdates) {
+    try {
+      const manifest = JSON.parse(update.manifestFragment);
+      const launchAssetKey: string | undefined = manifest.launchAsset?.storageKey;
+      const requestedUpdateId: string = update.id;
+      if (!launchAssetKey || !requestedUpdateId) {
+        continue;
+      }
+      toPrewarm.push({
+        update,
+        launchAssetKey,
+      });
+    } catch {
+      // ignore updates without launch assets
+    }
+  }
+
+  await Promise.allSettled(
+    toPrewarm.map(async ({ update, launchAssetKey }) => {
+      try {
+        // Check to see if there's a second most recent update so we can pre-emptively generate a patch for it
```

Comment:
I'm having a hard time wrapping my head around whether this needs to take into account per-update rollouts? I _think_ the answer is no, since the control will by definition be the second-most-recent update, but I may be forgetting something.